### PR TITLE
WIP: Encode/decode byte arrays

### LINKED ISSUE #4: Use package to base62 encode binary data
I'm interested in base62 encoding the bytes digest of a hash: https://github.com/greenelab/deep-review/pull/298.

It looks like `base62.encode()` takes an`int` not a `bytes` object. I assume the solution is to convert the bytes object to an `int`. @suminb do you have a suggestion for the best method to use?

### PR DESCRIPTION
**Pull request by suminb**
This is a pull request to resolve #4. I'm sure there is still room for improvement and some edge cases might not be handled correctly.

It would be greatly appreciated if @dhimmel could review this PR. Thanks.

### CODE REVIEW

File: base62.py [hunk 1]
```diff
@@ -27,6 +43,17 @@ def encode(n):
 
     return ''.join(s)
 
+
+def encodebytes(s):
+    """Encode a bytestring into a bytes object containing multiple lines
```
Comment (dhimmel) [inline base62.py]: In Python 3 `encode` and therefore `encodebytes` returns a str rather than bytes.

File: base62.py [hunk 2]
```diff
@@ -40,6 +67,25 @@ def decode(b):
...
+    _check_bytes_type(s)
+
+    decoded = decode(s.decode(DEFAULT_ENCODING))
+    buf = []
+    while decoded > 0:
+        buf.append(decoded & 0xff)
+        decoded = int(decoded / 256)
+    buf.reverse()
+
+    return bytearray(buf)
```
Comment (dhimmel) [inline base62.py]: I think this should return a `bytes` object rather than a `bytearray`. [In python 3](http://stackoverflow.com/a/5901849/4651668), `bytes` will be immutable like `str`. Also `bytes` support the `.decode()` which will be useful for getting text from decoded content...

```python
>>> base62_encoded = encodebytes('§'.encode())
>>> base62_encoded
'Cxj'
>>> bytes(decodebytes(base62_encoded.encode())).decode()
'§'
```

**Comment by dhimmel** [comment]
So I think we should determine the optimal types from a user perspective for `encodebytes` and `decodebytes` to consume and return. The difficulty is that Python's behavior regarding these types depends on 2 versus 3. Therefore, the proper type in Python 3 is not necessarily the proper type in Python 2.

Here's the API I imagine for Python 3: `encodebytes` should accept a [binary sequence type](https://docs.python.org/3/library/stdtypes.html#binary-sequence-types-bytes-bytearray-memoryview). `encodebytes` should return a `str`. For reciprocity, `decodebytes` should accept a `str` (currently requires a bytes-like object) and return a `bytes` object (currently returns a `bytearray` which is mutable and has no `decode` method).

If you want Python 2 to follow the same principles, `encodebytes` should consume a binary type (which could be `str` or `bytearray`). To behave more like python 3, `encodebytes` (and hence `encode`) should return a `unicode` string. Alternatively, you could return a `str` (as occurs now)... there's no perfect solution given's Python 2's conflation of str and bytes. Thus, `decodebytes` should consume either `str` or `unicode` and return a `str` (bytes).

@suminb what do you think?

**Comment by suminb** [comment]
@dhimmel Thanks for your detailed review. If there's a conflict between Python 2 and 3 in terms of type consistency or behavior, I'd go with Python 3's way. That's where future lies, right? ;-) I'll look into this when I get off work.

File: base62.py [hunk 3]
```diff
@@ -40,6 +72,25 @@ def decode(b):
...
+    _check_bytes_type(s)
+
+    decoded = decode(s.decode(DEFAULT_ENCODING))
+    buf = []
+    while decoded > 0:
+        buf.append(decoded & 0xff)
+        decoded //= 256
+    buf.reverse()
+
+    return bytearray(buf)
```
Comment (dhimmel) [inline base62.py]: I think you should wrap the `bytearray` in `bytes()` before returning. This will result in the desired behaviors for Python 2 & 3.

File: base62.py [hunk 4]
```diff
@@ -40,6 +72,25 @@ def decode(b):
 
     return v
 
+
+def decodebytes(s):
+    """Decodes a bytestring of base62 data into a bytes object.
+
+    :param s: A bytestring to be decoded in base62
+    :rtype: bytearray
```
Comment (dhimmel) [inline base62.py]: Should return `bytes` not `bytearray`.

File: base62.py [hunk 5]
```diff
@@ -40,6 +72,25 @@ def decode(b):
...
+    """
+    _check_bytes_type(s)
+
+    decoded = decode(s.decode(DEFAULT_ENCODING))
+    buf = []
```
Comment (dhimmel) [inline base62.py]: Rather than creating a list that is later converted to a `bytearray`, just declare the bytearray here. `bytearray` support `append` and `reverse`.

File: base62.py [hunk 6]
```diff
@@ -9,6 +9,27 @@
...
+
+
+def bytes_to_int(s, byteorder='big', signed=False):
+    """Converts a byte array to an integer value. Python 3 comes with a
+    built-in function to do this, but we would like to keep our code Python 2
+    compatible.
+    """
+    try:
+        return int.from_bytes(s, byteorder, signed=signed)
+    except AttributeError:
```
Comment (dhimmel) [inline base62.py]: Nice solution!

File: base62.py [hunk 7]
```diff
@@ -40,6 +72,25 @@ def decode(b):
 
     return v
 
+
+def decodebytes(s):
+    """Decodes a bytestring of base62 data into a bytes object.
+
+    :param s: A bytestring to be decoded in base62
```
Comment (dhimmel) [inline base62.py]: Should accept a `str` (i.e. bytes in py2, text in py3) rather than `bytestring`.

File: tests/test_basic.py [hunk 8]
```diff
@@ -1,8 +1,37 @@
...
+]
+
+
 def test_basic():
     assert base62.encode(0) == '0'
     assert base62.decode('0') == 0
 
     assert base62.encode(34441886726) == 'base62'
     assert base62.decode('base62') == 34441886726
+
```
Comment (dhimmel) [inline tests/test_basic.py]: I think a roundtrip test would be good.

For example, if you pass the following test in 2 and 3, I think you will have addressed my issues.

```python
input_bytes = b'my bytes to encode'
base62_encoded = base62.encodebytes(input_bytes)
assert isinstance(base62_encoded, str)
output_bytes = base62.decodebytes(base62_encoded)
assert isinstance(output_bytes, bytes)
assert input_bytes == output_bytes
```

File: tests/test_basic.py [hunk 9]
```diff
@@ -1,8 +1,46 @@
...
+@pytest.mark.parametrize('b, i', bytes_int_pairs)
+def test_encodebytes(b, i):
+    assert base62.encodebytes(b) == base62.encode(i)
+
+
+@pytest.mark.parametrize('s', ['0', '1', 'a', 'z', 'ykzvd7ga'])
+def test_decodebytes(s):
+    assert base62.bytes_to_int(base62.decodebytes(s)) == base62.decode(s)
+
+
```
Comment (dhimmel) [inline tests/test_basic.py]: Perhaps it'd make sense to `parametrize` this so we can test some more corner cases:

```python
@pytest.mark.parametrize('input_bytes', [b'', b'0', b'bytes to encode', b'\x01\x00\x80'])
```

**Comment by dhimmel** [comment]
@suminb after merge, did you want to release a new version on PyPI?

I'm excited to add this package to `greenelab/deep-review` (as discussed in https://github.com/greenelab/deep-review/pull/298#issuecomment-292722054)!

**Comment by suminb** [comment]
> @suminb after merge, did you want to release a new version on PyPI?

Yes, I will do that. Thanks for your review comments. I really appreciate it.

**Comment by suminb** [comment]
`pybase-0.2.0` has been released: <https://pypi.python.org/pypi/pybase62/0.2.0>
